bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/obsrecord/VisitCalculator.scala

```diff
@@ -0,0 +1,148 @@
+import edu.gemini.skycalc.TwilightBoundType.NAUTICAL
+import edu.gemini.skycalc.{Interval, ObservingNight, Union}
+import edu.gemini.spModel.core.{Semester, Site}
+import edu.gemini.spModel.dataset.{ DatasetLabel, DatasetQaState }
+import edu.gemini.spModel.dataset.DatasetQaState._
+import edu.gemini.spModel.event.{EndDatasetEvent, ObsExecEvent, OverlapEvent, StartDatasetEvent}
+import edu.gemini.spModel.obsclass.ObsClass
+import edu.gemini.spModel.syntax.all._
+import edu.gemini.spModel.time.ChargeClass
+import edu.gemini.spModel.time.ChargeClass._
+
+import scala.collection.JavaConverters._
+import scalaz.Scalaz._
+import scalaz._
+
+/**
+ * Calculates a VisitTimes for a sequence of events corresponding to a visit.
+ * VisitTimes describes how to charge time for the visit.
+ */
+private[obsrecord] sealed trait VisitCalculator {
+
+  /**
+   * Returns the Instant at which the calculator is first considered valid.
+   * Calculators are valid starting at a specific time per site.  The
+   * calculator that applies is looked up using the timestamp of the first
+   * event in the sequence and applies for the entire sequence thereafter.
+   */
+  def validAt(site: Site): Instant
+
+  /**
+   * Calculates the VisitTimes for the given sequence of events.  The QA
+   * obs class lookup functions are used for determining how to charge each
+   * individual dataset.
+   */
+  def calc(
+    events: VisitEvents,
+    qa:     DatasetLabel => DatasetQaState,
+    oc:     DatasetLabel => ObsClass
+  ): VisitTimes
+
+}
+
+/**
+ * Time accounting calculation based on observation events.
+ */
+private[obsrecord] object VisitCalculator {
+
+  /**
+   * The primordial visit calculator.  Adapts the old Java version to the
+   * VisitCalculator trait.
+   */
+  case object Primordial extends VisitCalculator {
+
+    implicit class Function1Ops[A, B](f: Function[A, B]) {
+      def asJava: java.util.function.Function[A, B] =
+        new java.util.function.Function[A, B] {
+          override def apply(a: A): B = f(a)
+        }
+    }
+
+    // This was the initial version.
+    override def validAt(s: Site): Instant =
+      Instant.MIN
+
+    override def calc(
+      events: VisitEvents,
+      qa:     DatasetLabel => DatasetQaState,
+      oc:     DatasetLabel => ObsClass
+    ): VisitTimes =
+      PrimordialVisitCalculator.instance.calc(events.sorted.toList.asJava, qa.asJava, oc.asJava)
+
+  }
+
+  /**
+   * The 2018B+ calculator.  The main change from the previous time accounting
+   * result is that visits without a passing dataset are not charged at all.  It
+   * also simplifies and fixes a number of bugs with the original algorithm:
+   *
+   * <ul>
+   *   <li>Charges for the portion of a dataset after start and before overlap</li>
+   *   <li>Handles start/end dataset pairs with intervening events correctly</li>
+   *   <li>Doesn't drop time if the first event is start dataset</li>
+   * </ul>
+   */
+  case object Update2018B extends VisitCalculator {
```

Comment:
Actually, now that I think about it.  2018B is coming right up.  This won't go into effect until 2019A.  Testing will be difficult.  I may need to set the start time to 2018B anyway so they can evaluate the impact it would have using 2018B data during the November testing period.